provider/pkg/provider/provider.go

```diff
@@ -465,7 +471,9 @@ func (k *commandProvider) Cancel(context.Context, *pbempty.Empty) (*pbempty.Empt
 	return &pbempty.Empty{}, nil
 }
 
-func (k *commandProvider) addContext(c context.Context) context.Context {
+func (k *commandProvider) addContext(c context.Context, mutex *sync.Mutex) context.Context {
+	mutex.Lock()
+	defer mutex.Unlock()
```

Comment:
I might be miss-understanding the go-specific mutex implementation, but it appears that we're creating a new mutex each time we're calling `addContext` at which point none of these would block each other from modifying `k.cancelFuncs` at the time time. I would have expected to see the mutex object initialized on the `commandProvider` (`k`) struct so we have a single instance which all threads would lock against.

Reply:
I've changed this around now